contracts/Compass.vy

```diff
@@ -104,39 +140,63 @@ def __init__(_compass_id: bytes32, _event_id: uint256, _gravity_nonce:uint256, v
+    self.assert_authority(authority, gas_estimate)
+    assert not self.message_id_used[message_id], "Used Message_ID"
+    self.message_id_used[message_id] = True
+    # check if the supplied current validator set matches the saved checkpoint
+    assert self.last_checkpoint == self.make_checkpoint(consensus.valset), "Incorrect Checkpoint"
+    # signing data is keccak256 hash of abi_encoded logic_call(args, message_id, compass_id, deadline)
+    args_hash: bytes32 = keccak256(_abi_encode(new_compass, message_id, deadline, authority, gas_estimate, method_id=method_id("upgrade(address,uint256,uint256,address,uint256)")))
+    # check if enough validators signed args_hash
+    self.check_validator_signatures(consensus, args_hash)
+    # last fee reservation
+    self.reserve_security_refund(authority, gas_estimate)
+    self.assert_balance()
+    # Make call 
+    success, response = raw_call(
+        new_compass,
+        _abi_encode(self.funds, self.claimable_rewards, self.total_funds, self.total_claims, self.rewards_security_wallet, self.rewards_community_wallet, method_id=method_id("receive_books(HashMap[bytes32,uint256],HashMap[address,uint256],uint256,uint256,uint256,uint256)")),
+        value=self.balance
+        revert_on_failure=True
+        )
+    assert success, "failed to send books"
+    self.funds = empty(HashMap[bytes32, uint256])
+    self.claimable_rewards = empty(HashMap[address, uint256])
+    self.total_funds = 0
+    self.total_claims = 0
+    self.rewards_security_wallet = 0
+    self.rewards_community_wallet = 0
+    self.assert_balance()
+    log BooksSentEvent(_value, new_compass)
+    return response
 
-# Make a new checkpoint from the supplied validator set
-# A checkpoint is a hash of all relevant information about the valset. This is stored by the contract,
-# instead of storing the information directly. This saves on storage and gas.
-# The format of the checkpoint is:
-# keccak256 hash of abi_encoded checkpoint(validators[], powers[], valset_id, compass_id)
-# The validator powers must be decreasing or equal. This is important for checking the signatures on the
-# next valset, since it allows the caller to stop verifying signatures once a quorum of signatures have been verified.
-@internal
-@view
-def make_checkpoint(valset: Valset) -> bytes32:
-    return keccak256(_abi_encode(valset.validators, valset.powers, valset.valset_id, compass_id, method_id=method_id("checkpoint(address[],uint256[],uint256,bytes32)")))
+# Receives the internal book keeping state of the last version of compass, along with any funds
+# left on the old contract.
+# Can only be sent by the old version of compass.
+@external
+@payable
+@nonreentrant('lock')
+def receive_books(_funds:HashMap[bytes32, uint256], _claims:HashMap[address, uint256], _total_funds:uint256, _total_claims:uint256, _rewards_security_wallet:uint256, _rewards_community_wallet:uint256):
```

Comment:
Same as above, does it make sense to validate `msg.value > 0`? Or better yet, only apply it if the current funds are uninitialized.

Reply:
I see you arrived at my conclusion before me. Yes, 100% this makes sense.